### ISSUE TITLE
Less support in stylelint rules

### ISSUE THREAD

**Issue by ssivanatarajan** [stylelint/stylelint#5375]
> Clearly describe the bug

Some of the rules are not giving correct result for Less code
[demo](https://stylelint.io/demo#N4Igxg9gJgpiBcIB0BDAFASgATADoDsssABAIwgCdYKt58J8YBuAo4gBxSigEt8BzWgEYADOwAeLQiXJUYNeBRhQpAXwIgANCADOAT3wAXFOIQgANjB06t4BgDMe-M3mm4QFAK6Wd7+DlYiLHcUQwBaL0sw+jDPfABregB3fD8sQy8YTUCid1JzCDB46IgwmABbdkM9NIzPLJzguwKKErC+ADcUcx4oMIALGHFazOzpXLty8pgjNoqqmoR00cb3WDBzFApQngYw-MLimKhPdh6wUJgwsE8dQwhysPYKCHZ5Qx4rEfqxoKb1zbbD57A5FNonM48C6GK7PV7vT6+JYAbUaRDqMEaAF1fkE1jANlsdiCCmCYjp+pRDP0UPg+nC3hRqmEIB15BRel8lhjcRN7AxwvYUOUeOY9ODTudLtFhVz-DzViB+bMhSKxW0RdY+PwwvwZvIoWFVaLxfEYHokpQoN8GuMmvY4mBgfhrt0wG1Ot1ejbefbHc7XeZ3TE4jpOGBlCzGaFKD7FQ78E7di6eowtrrtrwZuEYvR8HdaVAtn1eEokww43b3Ga9PZttMwgCiQGYjxKlTaYZK393NNeCgjTBQp4lDKGyGEslUtyVlWQPhZX1+By+lshzoPfguj1rTOforjlYI3TtWEwwSeI4wDxqt28fPShCpTCwqF2u2mRFvHLlvu58dJShaUdBgSwnUoJF5VnHsHzKSpmR0CARwjO8JhiIYMgHECRUgAp81QppWy3L0SyQ-Irh0TYKWuB5piMSDf1tGCiO3XonggHQb2TV9wjbdgqS-SwCPcVsKCUfhvHTJJ+hvKxw0xPcmPvBl3nFCdEggFJhN0UCCXuVp2BAzwoFKQlrDaOINK0xTfXcECwP0p4jJMspLDonNSksqdtPsvTKDCao3gsydNOnfxUTtdFoL+AJIvvJx6CUNIIpimL3BuO4HlcipsyRNEgixfL1DtHFFTuDkBDaRgklTBSoL-GC4hvYKrLCxjbJAN9IiuBcGwuEC0ncAoknkdwOq678xwo+TX3sGEKEGzrzCSFA9A3TiBCiWqxsVCaomwqEIDw6IYBqvgrhQObRqWEJltWpEOtBYoNg409SG2CM4IWMJav2GB+SSm75xgNkFq0RUnuuAoNp1d6UE+6rfsu+bFu6Fa1p2udIZemH9g+nrTt+0h-soOqmjR+6wnKbwPh+87MZg7Hobe-Gzxm4mAbJ270fW7Utvp8GsdJYp4XwFn4YJs7GFmlGgYptaqZpng6cYBn70h0Xxc+sMJZl67-G5ymYf51XBcZ4WoxmLXpt1jnSdRu6MbN+9cP8wZxFdAageG66Otd1p3Z+mZ+GpRaKSpNWJkgKZsy+5kiZJwHwvyw2nd9IhXFSiYhgjKpkvymDHAoO4TruZR3ALrBSriiYEvtlFK4mO49EsVNwmj8pCzymuq6K7E-do2PpNknXPr4TjYAdnnI6aDL7keFSmXFeZ4-Ov7OfzuLU4e-LM6z9wc5gPOG57iZkfka5bnnp4XkZW9naz+0eGLnMrBhXdK+rx-3DrpOsBSx+5MrqtA7tmGeqUf75k5GePmVwibCwrj3QqcVip-C-veJsQJuKkFpDqUeF1gGLUYKDGe+JCRYJBLgtmttE5cyWtPB+-wCSAmJC6SGB1XYukRmvc+YMDb0MptTcwtNtqMLISwgM7CKiHWOvgvWfDyaO15ptOBAsOqYNYfsC2HCjp7DkXbP+IkQa+0VBoyR2jYEqyuJ3D2ZjkwMSEOo5hzZsEW0wqKU8Oi8JT3uqQkAdi9icJOlLAhst+Hyw3EIkRajTHOIoS6IJcjeE+IVsbVRpsnHkM0Yk9mtCiHGLBpkiR3EV7igThvE+ECBFp13pXA+4hc5dkqYAkIwDGxxNYeAtKSpn4l0YGXD+SD07xX4IlMmADv6dTaaAowXSYLj2gWkqxWjDiIMfsg1KqCCodQTOWBJtEBzcOlskuWSjFbCOVqInZ-puIdywjNE54SzlLKufGG5gSDnUM+gYuhxCTFzl2QGGxccynnQYiIa5iYWyyk9nQn2hS3lQu4pwJQRhBggQ3Ecq4CzJ6nJ5uc6JGTEV7KeGuNFVgrBfOxVA3F-C-mtBeTEgF7yXTDxhEkwheLfFiJAHwWARhOlLAAEwdUsAIakYQABe8hSgThvNpYFpSrEOI6n2HgA57BDkMCOK4OTdaPMUQw1Vyh1WDmHKOPV3y8lA3pX4tVGqtU6qmrCxa8K7Umodea2EZLqQUvWjNHFvyCnuv7Ga7Vo5tgCCuPCIE-kOVhMNdy41obNVeoiLg6N0ZHL6OtU8o1ip7VhAAI71AoKCkudzglI05XmwRSsrEhtNSW+Q5b26fPjfrRNqTLGvLnIW5tZa6YVvbbkipdLg08vQkdMIbK5IS20jETUuMZh9AgPYM8SEKAI0JgLeqSk0KeHKMTVolguCnmlS8FJO9-yHuPW0dxqYdQXo4tpRezJ+pws0v8mC3UQXLJ+ZvKpESCWXKZalPeVTD7H2TqfJ+L9S7vzWVndBVTf7jMbkA+aNEY6zJAJ-PuJUOq+XAq0UIGQeCkE8C+OGRQYCGH9brQN+SSE8uI45MjHJKMvhjTGBlDya2EQnUR3SJGeLka45m+QvGqXr3ruOljwmHL+WjqQPg0mO0KO3n4tjymHiqYXNm0dcmu3Xpgjp1osAdDHiLLMFTanHK5j0fJHyInHI9GHVMQ5O7jkCa06x1z-l3Nts8zJgDNqhNlQC60RV8FQX9LSBCyLSmDLOVMlRDcH7XVfoRXOczTkYDGTSygcyKLszospXIpj4WFNJb8ilgrLldLuRdd7bL2mov5cK9lZrQT6DGGdItEynhyLteSwFPQQVMutZGjlxqYt27Fc-TNvx256hhBrJaKgLX+Fup5at1RHnO5Vp4b56pkT629pgvtodwWjsaavTAlRDa9vdDW0F7Dd2jOGOBjVuc133sxe+rVcFgRiqqBAKoIAA)
> Which rule, if any, is the bug related to?

"at-rule-no-unknown  and declaration-block-no-duplicate-properties"

> What code is needed to reproduce the bug?

```less
.a() {
  @border :none;
  @padding :10px;
  @border :red;
}

```

> What stylelint configuration is needed to reproduce the bug?

e.g.

```json
{
  "rules": {
    "declaration-block-no-duplicate-properties": [
      true
    ],
 "at-rule-no-unknown": true
  }
}
```

> Which version of stylelint are you using?

latest version

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

demo in stylelint website

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

yes it is related to less

> What did you expect to happen?
```
`at-rule-no-unknown`

    No errors should be reported because these are less variables
`declaration-block-no-duplicate-properties`
   
   It should report @border as duplicated.
```
> What actually happened (e.g. what warnings or errors did you get)?
```
`at-rule-no-unknown`

All these Less variables are reported as "unknown at rule"

`declaration-block-no-duplicate-properties`
duplicate variables are not reported

```

**Comment by jeddy3** [comment]
@ssivanatarajan Thanks for the report.

`at-rule-no-unknown`:

It seems [the parser](https://github.com/shellscape/postcss-less) expects the following whitespace:

```
@border: none;
```

[See demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIB0BDAFASgATADoDsssABAIwgCdYL4t8J8YBuAo4gBxSigEt8BzWgEYADOwAeLQiXJUYNLBRhQpAXwIgANCADOAT3wAXFOIQgANjB06t4BgDMe-M3mm4QFAK6Wd72q6IidxRDAFovS1D6UM98AGt6AHd8PyxDLxhNVkD3UnMIMDioiFCYAFt2Qz1U9M9M7KC7fIpi0L4ANxRzHihQgAsYcRqMrOlGyDKymCNW8srqhDSRhqx3WDBzFAoQngZQvIKi6KhPdm6wEJhQsE8dQwgy0PYKCHZ5Qx4rYbrRwNWQdabbYfPYHQqtE5nHgXQxXZ6vd6fXyLADaKyItRgKwAur8cgCYBstjtQflwdEdH1KIY+ih8L14W8KFVQhB2vIKD0votMXjGvYGGF7CgyjxzHoIadzpcoiLubReSt3AKZsLReLWqLrHx+KF+NN5NDQmqxRK4jA9IlKFBvvUxv97LEwCD8NcumBWh0uj1bXyHU6XW7zB7orEdJwwMpWUyQpRfUqQI78M7dq7uowtnrtrxpmFovR8Hc6VAtr1eEoUwx4-b3Oa9PZtlNQoDiYHojwKtS6YZq393FNeChjTAQp4lLKm6H4kkUjzljWQPg5b1+JzelsRzpPfhOt0bXOfgnjlZI-SdaFw4SeI4wDwqr38ccpdCZSE2p3meFvPKloeF0+oRhK4dBgSxnUoZEFXnPtFxKOYWR0CAx0jB9GmiQZ0iHEDRUgfJC1Q-52x3b0yyQvJgM2SlrgeKYjEg387Rgojdx6J4IB0O9U1CN8O3Yakv0sAj3HbCglH4bxM0SPo7ysCMsQPRj8UZd4JSnBIIGSITdFAwl7hadgQM8KASiJaxWlidTNIUv13BAsC9KeQzjNKSxaLzEoLJnLS7N0yhQiqN5zOnDTZ1oNF7QxaC-hwdF8SceglFScLoui9wbjuB4XPKXNkViohsVi9R7VxBM7k5ARWkYRJ03kqC-xg2I7yCyzQoYmyQDfCIriXJsLhA1J3HyRJ5HcdrOu-CdgLk7j7FhCgBo68xEhQPQtw4gRIhq0aE3GyJsOhCA8KiGBqr4K4UFmkbFmCJaVuRdqwSKDZ2PPUhtkjUoKhZGr9hgAVEuuxcYHZeatATR7rnydbdTelAPqqn6LrmhaumW1btoXCHnuh-Z3u6k6ftIP7KFq-5Ubu0Iym8D5Qi2sHMbJJ6odevGL2mon-tJm60bWnVNrOjGYIhhF8BZuH8dOxgZuRwHydWynqZ4WmBfpoXGejaYxY+8Nxelq7aG5inof5xhBfxYW3lFirYe19niYBg3Fp5s3xkOvyBnEN1+sBoarva3D3cGWnpn4GkFspakXf+CY3M++ZlaljmSaS2LDfR1XooCFL-kGSNKhTiLUsTHgKDuY67mUdw8qwErC8aeLk9RavGjuPRLHTMIY+LXK65rwqcX9mjc1CKSZJ1j6+A42AUdu9P-due5HmU5kJXgiVCft0nkqLuX7tirOUvcXOYHzpve8aJH5GuBfMuX+8M+zh0S7LxgK-3ava8f9wG4drBt6-jql0Wgx1zFHQ+IBJ5cgvHzK4hNGZV17gVQuRU-if3xC2YEXFSB0l1OPc6QCFqMBBlHNYhIgQkldNgiqeDfqcxns7B+-wMEUP2OrfaAdXQIzOnrUGjtd4K3MDTOm7VmGBghuwt2roaGX14WTWevMNqwJViIshrYsFsPKAdI6NCk6-2EsDP2CZRHqMONAxRCcrhlBMM2VRmCGD0SECookdjKHq0wmKc8Ei8L0LuiQgkziWEcOOpLfBMs+HyIEUI5RRjbGBMkWzXWMifHy2Nko02jDSEBMDEEnRm9CEGNBk48hgY14WNoY3MKqcna+MYUQA+Rdj6n0qefMmQCbFZNTGAoujhS55isLCd+iC-QwR-lvZurS5rUUmKA2pRdIGwDMfwE2VxHoIMfkglKKDAhoP5AGLiXdgmIwIbLCJVNBFK2EQmJMlZXQHOkcc8JPNFnLL8dc7JNEsJ2zoYDIhhiFxvK4lYz2pSar0REO1AFewepXD6lzCwGk-kwUha6TgSgjADBAluLhUt5lwv4WcqJ6SIV7L2Ki3MGKrAJInoWLk+SQbPLSfJYlyZAyj1hPcsJciGHtT4LAIwFDUgACZ2qWAEDSUIAAveQJQpx3i0kCuO30zoOPagOHgQ57AjkMGOGF8SOX6y5TU1Vyh1XDlHOOHJXyKmEQKX4tVGqtU6sml7OFvtCkJntWa7V44yXoqsJSmhuK6WIvxJ6zV5qrjbAEFcBEwI-L6tkWnPeC4w2OvHFG-UGs40tFyd8x5RqPUmqHAARzqBQdePAy4HOxaEg1SbIkXOiSmotoRS3yArVWj5VLa2JuqSkmBFi7UtrbeW5WnbJifN1rouFvz3X-jgodEe0l2VyS0tELUONpi9AgPYC8SEKDwwJirOqik0KeDKETFolguDnilS8ZJyamLnsva0dx6ZdR3vYlpO+EpYULTdX4rqiqK2JzyWfcB+LFaDtmfUmCjSezgcfk-Xp5cBlrOzjsuZ-AEpjJacENpICjDoZShs6KWz8rtR8uBFoIR0g8FIJ4WEuM4bmkMGtaaQafm2oydpeyflaOcgY0x2NsYc3TSSVx4hPGqMOQE-RxjMaYwOVzda-RUnKM6Wo1M0gfBRPdp4Q+vxMm-ITB00uZTVq9F9sffiYzLRYA6FPCWGYpndMOXzHsPB3lNMOW6OOqxhzuESfzXPUqPm-J+c7l2lTVnZ1GfCy0BVILlWpHBWFvj+knImUoluP9PsEVzpgnZxyMAjLZZQGZX1NJ-Xsd1pxx2cXpMJZK2VrKsc8uOwA01jLLXnI6Xa-E+gxgXQLWMp4ci8WesBRhRV11BW-GNU7rN-982eO7jqKEOsVoqAupW8NQr+J1tKP80OGtBmTlPIJY2olCYjtjqixO-TwXDX9vMZchcd3ItTICzFmd3H2qfcrWEJLX0QM-hENkIqqgQCqCAA).

I'll close as a duplicate of https://github.com/stylelint/stylelint/issues/5210 as it's closely related.

Not many people use Less these days, so you'll need to contribute [to the parser upstream](https://github.com/shellscape/postcss-less) if you'd like to see issues like these fixed.

`declaration-block-no-duplicate-properties`:

This rule is behaving as expected as `@border: none` is an at-variable and not a property.

As the built-in rules are geared towards standard CSS, you'll need to write a [plugin](https://stylelint.io/developer-guide/plugins) for Less constructs. If you intend to continue using Less, then I suggest creating a `stylelint-less` plugin pack like [`stylelint-scss`](https://github.com/kristerkari/stylelint-scss). You can then create a `less/no-duplicate-at-variables` rule within it, like the equivalent [`no-duplicate-dollar-variables`](https://github.com/kristerkari/stylelint-scss/blob/master/src/rules/no-duplicate-dollar-variables/README.md) one in the [`stylelint-scss`](https://github.com/kristerkari/stylelint-scss) plugin pack.

If you do create `stylelint-less`, I suggest you also create a `stylelint-config-less-recommended` sharable config that bundles and configures it, like the equivalent [`stylelint-config-recommended-scss`](https://github.com/kristerkari/stylelint-config-recommended-scss) one. In the [upcoming major release of stylelint](https://github.com/stylelint/stylelint/issues/5205), the Less parser will be removed from stylelint (see https://github.com/stylelint/stylelint/issues/5289#issuecomment-840772900) and a config like `stylelint-config-less-recommended` should bundle the syntax (https://github.com/stylelint/stylelint/issues/5205#issuecomment-856669471).

Please let us know if you create these plugin pack and configs, so that we can point users towards them in our `14.0.0` migration guide.

**Comment by ssivanatarajan** [comment]
@jeddy3  I  will create the plugin for less .

**Comment by ssivanatarajan** [comment]
@jeddy3  I have created a [plugin](https://github.com/ssivanatarajan/stylelint-less) for Less  and also created sharable config for less [stylelint-config-recommeded-less](https://github.com/ssivanatarajan/stylelint-config-recommended-less). 

please have a look it and let me know anything needs to be changed. 

As of now added 3 rules in `stylelint-less` plugin , will add more rules. 

Thanks !!

**Comment by JounQin** [comment]
@jeddy3 

```less
@directions: {
  t: top;
  r: right;
  b: bottom;
  l: left;
};
```

Object-type variables reports `property-no-unknown` error, is expected or should be supported by `stylelint-less` instead?

It is said:

> This rule ignores:
>
> variables ($sass, @less, --custom-property)
> vendor-prefixed properties (e.g., -moz-align-self, -webkit-align-self)

But actually not ignored correctly.

**Comment by jeddy3** [comment]
@JounQin That's a bug in Stylelint that'll be fixed in `14.0.0` - see https://github.com/stylelint/stylelint/issues/4834.
